[material-ui] Add `storageManager` prop to `ThemeProvider` (#45136)
mui/material-ui: docs/data/material/customization/dark-mode/dark-mode.md

To use a different storage manager, create a custom function with this signature:

```ts
type Unsubscribe = () => void;

function storageManager(params: { key: string }): {
  get: (defaultValue: any) => any;
  set: (value: any) => void;
  subscribe: (handler: (value: any) => void) => Unsubscribe;
};
```

Then pass it to the `storageManager` prop of the `ThemeProvider` component:

```tsx
import { ThemeProvider, createTheme } from '@mui/material/styles';
import type { StorageManager } from '@mui/material/styles';

const theme = createTheme({
  colorSchemes: {
    dark: true,
  },
});

function storageManager(params): StorageManager {
  return {
    get: (defaultValue) => {
      // Your implementation
    },
    set: (value) => {
      // Your implementation
    },
    subscribe: (handler) => {
      // Your implementation
      return () => {
        // cleanup
      };
    },
  };
}

function App() {
  return (
    <ThemeProvider theme={theme} storageManager={storageManager}>
      ...
    </ThemeProvider>
  );
}
```

:::warning

> If you are using the `InitColorSchemeScript` component to [prevent SSR flicker](/material-ui/customization/css-theme-variables/configuration/#preventing-ssr-flickering), you have to include the `localStorage` implementation in your custom storage manager.

Reviewer: ```suggestion
If you are using the `InitColorSchemeScript` component to [prevent SSR flickering](/material-ui/customization/css-theme-variables/configuration/#preventing-ssr-flickering), you have to include the `localStorage` implementation in your custom storage manager.
```

:::

### Disable storage

To disable the storage manager, pass `null` to the `storageManager` prop:

```tsx
<ThemeProvider theme={theme} storageManager={null}>
  ...
</ThemeProvider>
```

:::warning

> Disabling the storage manager will reset the app to its default mode when users refresh the page.

Reviewer: ```suggestion
Disabling the storage manager will cause the app to reset to its default mode whenever the user refreshes the page.
```

But if you want to be able to toggle between modes manually, avoiding the flicker requires a combination of CSS variables and the `InitColorSchemeScript` component.
Check out the [Preventing SSR flicker](/material-ui/customization/css-theme-variables/configuration/#preventing-ssr-flickering) section for more details.

> ## Storage manager

Reviewer: Is this the best location for this info? Or would it make sense as an h3 under the System preference header? The fact that it refers back to the Built-in support section for context makes me think they should be located near each other. But if you have a strong opinion about this placement then I'm sure it's fine.

Author: I think it should be h2 because it's a different topic similar to `Toggling color mode` and the reason I put this section at the end is because I expect only a few projects looking for this docs.

Do you think it should be moved next to `Toggling color mode`?

Reviewer: Got it, in that case yes I think placing it near Toggling color mode makes sense. If you read these two sections back-to-back it does have a logical flow, even if this new addition is somewhat of an edge case.

> The default approach of the [built-in support](#built-in-support) uses the browser's `localStorage` API to store the mode and color scheme preference.

Reviewer: ```suggestion
By default, the [built-in support](#built-in-support) for color schemes uses the browser's `localStorage` API to store the user's mode and scheme preference.
```

To use a different storage manager, create a custom function with this signature:

```ts
type Unsubscribe = () => void;

function storageManager(params: { key: string }): {
  get: (defaultValue: any) => any;
  set: (value: any) => void;
  subscribe: (handler: (value: any) => void) => Unsubscribe;
};
```

Then pass it to the `storageManager` prop of the `ThemeProvider` component:

```tsx
import { ThemeProvider, createTheme } from '@mui/material/styles';
import type { StorageManager } from '@mui/material/styles';

const theme = createTheme({
  colorSchemes: {
    dark: true,
  },
});

> function storageManager(): StorageManager {

Author: ```suggestion
function storageManager(params): StorageManager {
```

  return {
    get: (defaultValue) => {
      // Your implementation
    },
    set: (value) => {
      // Your implementation
    },
    subscribe: (handler) => {
      // Your implementation
      return () => {
        // cleanup
      };
    },
  };
}

function App() {
  return (
    <ThemeProvider theme={theme} storageManager={storageManager}>
      ...
    </ThemeProvider>
  );
}

> ```

Reviewer 2: We should also include an example of how to turn it off entirely with `storageManager={null}`, including the consequences of doing so.
